Handle minibuffer completion table, predicate and metadata

clemera opened the pull request:
This will allow us to access the metadata for purposes as discussed in  #15 and #82 and allows other packages and commands to access those variables (see #94). With this patch we can retrieve the metadata like:

    (completion-metadata input
                         minibuffer-completion-table
                         minibuffer-completion-predicate)

We could also decide to only pass the metadata instead (which would probably reduce garbage collection when exiting the minibuffer I guess). But the metadata can depend on the current input and the predicate passed to the completion table. Having metadata dependent on the input or predicate sounds like a crazy idea and I don't know if it actually happens in practice.

clemera commented:
This PR fails with CI for emacs-master but I can't check why.

raxod502 commented:
![image](https://user-images.githubusercontent.com/6559064/82756388-34391480-9d97-11ea-9a01-42bcf58440a2.png)

You should also be able to test locally, `make docker VERSION=master`.

clemera commented:
Thanks, it's fixed now.

raxod502 commented:
See https://github.com/raxod502/selectrum/issues/94#issuecomment-633641190.

raxod502 commented on selectrum.el:
```diff
@@ -1082,7 +1084,12 @@ is very confusing.
+For MINIBUFFER-COMPLETION-TABLE and
+MINIBUFFER-COMPLETION-PREDICATE see `minibuffer-completion-table'
+and `minibuffer-completion-predicate'. They are used for internal
+purposes and compatibility to Emacs completion API."
```
This should probably explicitly mention that passing the keyword arguments will dynamically bind the corresponding variables, as otherwise you have to be extremely familiar with binding semantics and the implementation of `cl-defun` to guess that it has this effect.

clemera replied:
Good idea, have done it.

clemera commented:
I have added the change to the CHANGELOG. I also added an additional change I would like to propose that I haven't added to the CHANGELOG yet so we can discuss it first.

The [embark](https://github.com/oantolin/embark/) package has made a lot of progress and can be used with selectrum for occur and actions. Currently there is one last issue with selectrum which the last commit fixes: When calling an action embark calls the action and insert the last candidate into the minibuffer and exits. With default completion there is no extensive computation when inserting a candidate into the minibuffer so this works well. With selectrum the candidates get computed beforehand which can introduce a noticeable delay. 

With that embark users would be able to use the following setup and wouldn't get the delay anymore:

```elisp
(add-hook 'embark-setup-hook
          (defun embark-selectrum-handle-inject ()
            (when selectrum--active-p
              ;; use the injected cnadidate and make sure selectrum doesn't
              ;; update afterwards
              (setq selectrum--current-candidate-index -1)
              (setq selectrum--previous-input-string
                    (buffer-substring
                     selectrum--start-of-input-marker
                     selectrum--end-of-input-marker)))))
```

clemera commented:
I also added support for `display-sort-function` metadata now. You can test with:

```elisp
(completing-read "Check: "
                 (lambda (string pred action)
                   (if (eq action 'metadata)
                       (list
                        'metadata (cons 'display-sort-function
                                        (lambda (seq) (sort seq #'string<))))
                     (complete-with-action action '("c" "b" "a") string pred))))
```

clemera commented:
Now also added support for annotation metadata:

```elisp
(completing-read "Check: "
                 (lambda (string pred action)
                   (if (eq action 'metadata)
                       `(metadata
                         (annotation-function
                          . ,(lambda (str) (concat " annotation for " str))))
                     (complete-with-action action '("c" "b" "a") string pred))))
```

clemera commented:
This should be ready, if you are okay with the changes I will update the CHANGELOG.

clemera commented:
`selectrum-completion-in-region` can be refactored to make use of handling the metadata within `selectrum-read`, too. If you want I can do this in this PR or open another PR at some later point if you prefer.

clemera commented:
`selectrum-completion-in-region` could now be refactored to make use of handling the metadata within `selectrum-read`, too. If you want I can do this in this PR or open another PR for that at some later point if you prefer. Other than that this should be ready now, if you are okay with the changes I will finish updating the CHANGELOG.

clemera commented:
Thanks, I finished updating the CHANGELOG.

clemera commented:
Thanks, I finished updating the CHANGELOG.

clemera commented:
Thanks, I finished updating the CHANGELOG and also added the mentioned `selectrum-exhibit` function from above.

clemera commented:
Thanks, I finished updating the CHANGELOG and also added the mentioned `selectrum-exhibit` function from above.

raxod502 commented on CHANGELOG.md:
```diff
@@ -64,6 +64,19 @@ The format is based on [Keep a Changelog].
+  `annotation-function` or `display-sort-function` in its metadata,
+  Selectrum will use this information to annotate or sort the
+  candidates accordingly ([#82], [#95]).
+* One can trigger an update of Selectrums completions UI manually by
```
```suggestion
* One can trigger an update of Selectrum's completions UI manually by
```

raxod502 commented on selectrum.el:
```diff
@@ -475,6 +494,56 @@ This is used to implement `selectrum-repeat'.")
+    ""))
+
+(defun selectrum-exhibit ()
+  "Trigger an update of Selectrums completion UI."
```
```suggestion
  "Trigger an update of Selectrum's completion UI."
```

raxod502 commented on selectrum.el:
```diff
@@ -1082,7 +1165,14 @@ is very confusing.
+MINIBUFFER-COMPLETION-PREDICATE see `minibuffer-completion-table'
+and `minibuffer-completion-predicate'. They are used for internal
+purposes and compatibility to Emacs completion API. By passing
+theses as keyword arguments they will be dynamically bound as per
```
```suggestion
these as keyword arguments they will be dynamically bound as per
```

clemera commented:
Turned out that `completion-extra-properties` also need to be handled for `completing-read` (see `proced-send-signal` for an built-in example).

clemera commented:
Thanks for the feedback and review!
